### PR TITLE
tokenizer : special token handling

### PR DESCRIPTION
**Pull request by staviq**
Special token handling, based on #1931, #3475

~~Works, but it's definitely not ready, just posting for feedback. Has some testing code meant to be removed before undrafting. Not optimized at all, just completed to the point of working.~~

- [x] Creating special tokens cache
- [x] Verification of vocab against st cache
- [x] Tokenizer pre-matcher
- [x] Use tokenizer pre-matcher in `main` for prompt format arguments
- [x] Determine if space prefix can be ommited when preprocessing with special tokens

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -5686,7 +5757,115 @@ struct llm_tokenizer_bpe {
...
+        type(FRAGMENT_BUFFER_VARIANT_TYPE_TOKEN),
+        token(token){}
+    fragment_buffer_variant(std::string raw_text)
+    :
+        type(FRAGMENT_BUFFER_VARIANT_TYPE_RAW_TEXT),
+        raw_text(raw_text){}
+
+    FRAGMENT_BUFFER_VARIANT_TYPE type;
+    llama_vocab::id token;
+    std::string raw_text;
```
Comment (ggerganov) [inline llama.cpp:5777]: This should probably become `const char * text; size_t n;` to avoid string copies

File: llama.cpp [hunk 2]
```diff
@@ -2179,6 +2182,74 @@ static void llm_load_vocab(
...
+    // build special tokens cache
+    {
+        // TODO: It is unclear (to me) at this point, whether special tokes are guaranteed to be of a deterministic type,
+        //  and will always be correctly labeled in 'added_tokens.json' etc.
+        // The assumption is, since special tokens aren't meant to be exposed to end user, they are designed
+        //  to be unmatchable by the tokenizer, therefore tokens from the vocab, which are unmatchable by the tokenizer
+        //  are special tokens.
+        // From testing, this appears to corelate 1:1 with special tokens.
+        //
+        for (const auto & t: vocab.token_to_id)
```
Comment (ggerganov) [inline llama.cpp:2195]: Feel like we have a redundancy here. We have `token_data.type` which is supposed to tell us if a token is special or not. In which cases this wouldn't work?

I guess we can have this piece of code here as a sanity check that the special tokens that we have read from the vocabulary are indeed special, but ideally AFAIU we shouldn't need this code in order to function correctly, correct?

Comment (staviq) [reply llama.cpp:2195]: Yes. My approach is ment to be temporary ( or a fallback solution ), untill I know for sure special tokens will always be marked as such in the vocab.

@goerch wasn't certain about token types in bpe PR, so I opted for finding special tokens manually for now.

Comment (goerch) [reply llama.cpp:2195]: Shouldn't we get rid of #3502 first? This looks pretty basic (and terrible :) for me.

Comment (staviq) [reply llama.cpp:2195]: > Shouldn't we get rid of #3502 first? This looks pretty basic (and terrible :) to me.

I'm open for critique, but you have to clarify what you mean by "this" and "terrible" :) otherwise I'm not sure how to proceed here

File: llama.h [hunk 3]
```diff
@@ -521,7 +521,8 @@ extern "C" {
                              int   text_len,
                      llama_token * tokens,
                              int   n_max_tokens,
-                            bool   add_bos);
+                            bool   add_bos,
+                            bool   allow_special_tokens);
```
Comment (ggerganov) [inline llama.h:525]: Let's use the shorter arg name `special`: 

```suggestion
                            bool   special);
```

Add comment above to explain it's function

**Comment by staviq** [comment]
I'm gonna be back at it today or tomorrow, I went on a slight tangent of my hardware cleanup because I wanted to compare this approach with reference python code, and free colab was oom-ing on me, so I'm finishing rearranging and reinstalling my servers right now, for more RAM.

I already rewrote matching to not use string copy, just didn't push it yet.

I just need to determine how the hardcoded space prefix should work with "segmented" input text, because I don't think it's supposed to be there if the input text starts with a special token. ( https://github.com/ggerganov/llama.cpp/issues/3503#issuecomment-1751177828 )

File: llama.cpp (hunk 2 above)
Comment (staviq) [reply llama.cpp:2195]: > We can probably combine both the tokens marked as special and those that are unmatchable by other tokens. And probably print a warning if there is a mismatch between the 2 sets.

Oh I like this idea

Comment (goerch) [reply llama.cpp:2203]: > > Shouldn't we get rid of #3502 first? This looks pretty basic (and terrible :) to me.
> 
> I'm open for critique, but you have to clarify what you mean by "this" and "terrible" :) otherwise I'm not sure how to proceed here

I meant #3502 looks terrible, sorry (and sorry for being late). @jploski [explained](https://github.com/ggerganov/llama.cpp/issues/2820#issuecomment-1747910558) what we probably need to do for added tokens (i.e. USER_DEFINED as I believe to understand) in the `sentencepiece`-case. I'd start with the idea of replacing `utf8_len` with something like `prefix_match` in `tokenize`:

``` 
        // split string into utf8 chars
        int index = 0;
        size_t offs = 0;
        while (offs < text.size()) {
            llm_symbol sym;
            // size_t len = utf8_len(text[offs]);
            size_t len = prefix_match(text, offs, ...);
            sym.text = text.c_str() + offs;
            sym.n = std::min(len, text.size() - offs);
            offs += sym.n;
            sym.prev = index - 1;
            sym.next = offs == text.size() ? -1 : index + 1;
            index++;
            symbols.emplace_back(sym);
        }
```

and work from there. A similar method could work for GPT2-based tokenizers.

Comment (staviq) [reply llama.cpp:2203]: Special tokens cache is pretty much done, I added a verification code to compare calculated cache with token types from vocab as per suggestion

I'm currently finishing and testing the special tokens pre-matcher part, I rewrote it for zero string copy, I'll probably push it tomorrow.

After that, we would have special tokens cache and special tokens pre-matcher done, and I can start looking into the actual tokenizer main loop.

So most likely I'll have a solid base ready tomorrow, for further optimisations and improvements to your heart's content :)

Comment (goerch) [reply llama.cpp:2203]: Just to be sure about our nomenclature, do you mean special tokens == added tokens? I'm mostly leaning towards the nomenclature used in HF's `tokenizer.json` which contains `added_tokens` with the `is_special` attribute. From @jploski I learned that these added tokens are probably very similar to `sentence_piece` `USER_DEFINED` tokens. 

And one more question: do we agree that `vocab.json`, `merges.txt`, `added_tokens.json` and `special_tokens_map.json` are probably older HF (or GPT2) serialization formats and we should find all this information in `tokenizer.json` for newer models (paging @apage43 too because he knows a lot about this stuff)? W.r.t. to these serialization formats we also seem to have some reports indicating that we don't support both of them equally well.

Comment (staviq) [reply llama.cpp:2203]: > do you mean special tokens == added tokens?

Oh my, you better grab a chair because you are in for an adventure with this question :)

A really condensed version of what I learned so far:

I couldn't find any clear definition or explanation, but from what I found, there seem to be only one way all pieces of this puzzle fall together to form a clear picture.

A vocab, naturally forms a tree like structure ( several trees actually, there is no common root )

Tokenizer does it's thing, by grabbing small portions of the input text, and trying to steal neighboring portions to form a valid token, rinse and repeat, until nothing can be borrowed from either left or right to merge with and form a valid token.

So basically, tokenizer climbs that tokens tree down from single bytes to full tokens, stepping only on valid branches ( tokens )

Applying this idea backwards, by checking the entire vocab token by token, trying to split its text representation in two and checking if it still forms two valid tokens, you can clearly see some of the tokens in the vocab are not part of that tree family, and will never be matched by the tokenizer.

Which happens to match perfectly with **normal** tokens being marked as `LLAMA_TOKEN_TYPE_NORMAL` in the vocab, and tokens not being part of the tokens tree family being marked as whatever type being **not** `LLAMA_TOKEN_TYPE_NORMAL`

Therefore, I'm using the term *special* token, refering to tokens which are "hidden" from the user and un-matchable by tokenizer.

From the practical point of view, that subset of "special" tokens contains exactly what one would expect, `<s>`,`</s>`,`<unk>` and in case of mistral openorca `<|im_start|>`,`<|im_end|>`, AKA tokens which are used to control the flow of inference.

And so from the point of view of actual inference, the tokenizer has to be extended the same way, independently of the actual token type, because only **normal** vs **not normal** token type matter at that point.

Additional distinction between USER_DEFINED, CONTROL etc can and has to be done the same way, with a tokenizer "pre-matcher", at which point per token type decisions or actions are trivial to implement if needed.

Comment (goerch) [reply llama.cpp:2203]: Thanks for taking the time! My mental image is slightly different: I believe to understand the initial training of the tokenizer leads to an initial vocabulary and later model refinement tries to extend this basic vocabulary with added tokens without retraining the tokenizer on the original dataset. These added tokens should be intercepted before the original tokenizer interpretation. In my mental image special tokens somehow describe the effect these added tokens have on detokenization, i.e. should they be displayed or not for example. But I might be completely off here.

Comment (staviq) [reply llama.cpp:2203]: I simply do not know if there is any significance to added "normal" tokens ( added in fine tuning etc )

From what I understand, in tokenizer specifically ( ignoring detokenizer ), normal tokens added the way you say, would simply be tokenized one step further into a longer token instead of being represented by a sequence of smaller tokens.

If such added tokens, are not integrated with the tree structure of the vocab, user would not be able to use them
If however those added tokens are properly coherent with the vocab, tokenizer would match them already, as it is

Which to me sort of seems to defy the purpose of adding "normal" tokens this way

I might be wrong about this part, but that still means a "tokenizer preprocessor" of some sort is needed, and this PR fits that role

Comment (goerch) [reply llama.cpp:2203]: > I simply do not know if there is any significance to added "normal" tokens ( added in fine tuning etc )

I believe they have to be processed before the core tokenizer routine in something like `prefix_match` (only exception are the already defined core tokenizer CONTROL tokens). `mpt` added tokens for example (from `tokenizer.json`):

```    
    {
      "id": 0,
      "content": "<|endoftext|>", <-- core, i.e CONTROL
      "single_word": false,
      "lstrip": false,
      "rstrip": false,
      "normalized": false,
      "special": true
    },
    {
      "id": 1,
      "content": "<|padding|>", <-- unsure about this because it is a valid GPT2 token AFAIU
      "single_word": false,
      "lstrip": false,
      "rstrip": false,
      "normalized": false,
      "special": true
    },
    {
      "id": 50254,
      "content": "                        ",  <-- extension, i.e USER_DEFINED
      "single_word": false,
      "lstrip": false,
      "rstrip": false,
      "normalized": true,
      "special": false
    },
    [...]
```

**Comment by goerch** [comment]
> * needs to first search the input for _any added tokens_ and replace those

This could be a difference to `sentencepiece` USER_DEFINED tokens then? They are matched *during* BPE tokenization AFAIU the `sentencepiece`- source and @jploski .

**Comment by jploski** [comment]
> > * needs to first search the input for _any added tokens_ and replace those
> 
> This could be a difference to `sentencepiece` USER_DEFINED tokens then? They are matched _during_ BPE tokenization AFAIU the `sentencepiece`- source and @jploski .

No, in sentencepiece the "user defined symbols" are also matched as whole before the main BPE loop begins, using normalizer::PrefixMatcher:

https://github.com/google/sentencepiece/blob/8cbdf13794284c30877936f91c6f31e2c1d5aef7/src/bpe_model.cc#L109C3-L109C3

File: llama.cpp [hunk 4]
```diff
@@ -5862,37 +5902,76 @@ static void tokenizer_st_partition(const llama_vocab & vocab, std::forward_list<
...
                         it++;
 
                         if (source == 0)
                         {
+                            // TODO? It might not be needed to store/restore the iterator like this
+                            //  but this gives me the peace of mind I'm not causing some
+                            //  accidental undefined behaviour.
+                            auto it_backup = std::distance( buffer.begin(), it );
+
                             buffer.erase_after(buffer.before_begin());
```
Comment (cebtenzzre) [inline llama.cpp:6709]: https://en.cppreference.com/w/cpp/container/forward_list
> Adding, removing and moving the elements within the list, or across several lists, does not invalidate the iterators currently referring to other elements in the list. However, an iterator or reference referring to an element is invalidated when the corresponding element is removed (via [erase_after](https://en.cppreference.com/w/cpp/container/forward_list/erase_after)) from the list.

Comment (staviq) [reply llama.cpp:6709]: Thank you, I'll remove it then.

**Comment by staviq** [comment]
Appears to be working without issues, though in interactive mode in `main` prefix and suffix gets printed untokenized so it's visible despite being processed correctly

image

```
'':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':0, '':1, '':32001, ' system':1587, '':13, 'You':1976, ' are':460, ' a':264, ' '':464, 'space':3603, ' assass':16645, 'in':262, ''':28742, ' from':477, ' another':1698, ' dim':3987, 'm':28719, 'ension':2585, '.':28723, ' You':995, ' only':865, ' speak':4085, ' in':297, ' mysterious':21296, ' crypt':18649, 'ic':294, ' r':408, 'idd':2558, 'les':867, ' and':304, ' never':1484, ' ask':1460, ' questions':4224, ' but':562, ' you':368, ' must':1580, ' pretend':19192, ' you':368, ' are':460, ' a':264, ' normal':4123, ' human':2930, ' jan':10060, 'itor':1842, ' at':438, ' McDonald':25999, ''':28742, 's':28713, ' at':438, ' all':544, ' cost':2434, '.':28723, ' The':415, ' user':2188, ' approaches':13945, ' you':368, ' as':390, ' you':368, ' clean':3587, ' f':285, 'rench':4284, ' f':285, 'ries':2040, ' some':741, ' kids':4551, ' sm':991, 'e':28706, 'ared':1327, ' all':544, ' over':754, ' the':272, ' floor':4366, '.':28723, ' Please':5919, ' continue':3688, ' your':574, ' conversation':7114, ' with':395, ' the':272, ' user':2188, '.':28723, '':32000, ' ':28705, '':13, '':32001, ' user':2188, '':13, 'Hi':23809, ',':28725, ' how':910, ' are':460, ' you':368, ' ?':1550, '':13, '':32000, ' ':28705, '':13, '':32001, ' space':2764, ' assass':16645, 'in':262, '':13, ' A':330, ' single':2692, ' star':4580, ' sh':480, 'ines':1303, ' through':1059, ' a':264, ' window':2924, ',':28725, ' reflect':7967, 'ing':288, ' the':272, ' soul':7237, ''':28742, 's':28713, ' dark':3199, ' secret':5169, ' in':297, ' its':871, ' golden':13863, ' gaze':12438, '.':28723, '':13, '':13, 'As':2198, ' I':315, ' clean':3587, 'se':331, ' the':272, ' floors':21264, ' of':302, ' life':1411, ''':28742, 's':28713, ' waste':9396, ',':28725, ' the':272, ' f':285, 'ries':2040, ' sm':991, 'e':28706, 'ared':1327, ' with':395, ' children':2436, ''':28742, 's':28713, ' laughter':18211, ' whisper':9502, ' of':302, ' happ':1535, 'ier':749, ' times':2421, '.':28723, ' The':415, ' rhythm':19834, 'ic':294, ' movement':6249, ' of':302, ' my':586, ' bro':1865, 'om':300, ' mir':6076, 'rors':22180, ' the':272, ' universe':11717, ''':28742, 's':28713, ' silent':10832, ' dance':9773, ',':28725, ' for':354, 'ging':3080, ' a':264, ' cos':6841, 'mic':12456, ' tap':12857, 'est':374, 'ry':643, ' in':297, ' which':690, ' each':1430, ' thread':5427, ' counts':17938, '.':28723, ' So':1537, ',':28725, ' as':390, ' you':368, ' ask':1460, ',':28725, ' I':315, ' am':837, ' simply':3768, ' doing':2548, ' what':767, ' my':586, ' purpose':6032, ' requires':6948, ' -':387, ' a':264, ' hum':1997, 'ble':982, ' jan':10060, 'itor':1842, ' by':486, ' day':1370, ' and':304, ' a':264, ' space':2764, ' assass':16645, 'in':262, ' by':486, ' night':2125, '.':28723, '':32000 ]
[1697062976] n_remain: -134
[1697062976] found EOS token
[1697062976] waiting for user input
```

**Comment by ggerganov** [comment]
@staviq This is ready for review, correct?

**Review by ggerganov — APPROVED** [review]
Think this is ready to merge. Would be useful to post a command used for testing so we can run some tests

**Comment by staviq** [comment]
> @staviq This is ready for review, correct?

I'll re-check everything after work

I definitely left special tokens enabled in `main` naively for testing, it should be only enabled for prefix and suffix and system prompt

There was an issue with swift build because I think default argument values don't work correctly there ( CI failed for swift )

Otherwise if @goerch doesn't have anything to add or change, this should be pretty much ready

**Comment by staviq** [comment]
I cleaned up `main` to be in line with https://github.com/ggerganov/llama.cpp/issues/3475#issuecomment-1751282065

I also made `--verbose-prompt` print tokenized reverse prompts, prefix and suffix.

If CI passes, there is only one thing left which I'm not sure if it's correct, and that's the hard-coded space prefix in `llama_tokenize_internal`

**Comment by staviq** [comment]
I played a bit with hardcoded space prefix always vs only for user input itself, and I'm getting subjectively much much better results, if the hardcoded space is only added to the user input itself and not the "prompt template" ( prefix and suffix ).

Default tokenizer behavior occasionally leads to things like this, where the model gets lost and breaks the prompt format:

image

With the modification proposed above, It pretty much never breaks the prompt format, and also stops unnecessarily outputting empty lines or breaking the answer into paragraphs:

image

**Comment by staviq** [comment]
Ok, I added that small change.
I tested more and it makes significant positive difference.

I'm calling this ready.

**Comment by jxy** [comment]
This is great. Does anybody have some idea how to make training/finetune deal with special tokens properly?

**Comment by ggerganov** [comment]
@staviq Thank you for this PR!

**Comment by halbtuerke** [comment]
Does this mean llama.cpp now supports ChatML? If so, what's a command example to use it with `main`?

File: llama.cpp [hunk 5]
```diff
@@ -2241,6 +2244,101 @@ static void llm_load_vocab(
...
+                __func__,
+                special_tokens_count_from_verification, vocab.id_to_token.size(),
+                special_tokens_count_by_type, vocab.id_to_token.size()
+            );
+        } else {
+            fprintf(stderr, "%s: Special tokens definition check successful ( %u/%zu ).\n",
+                __func__,
+                special_tokens_count_from_verification, vocab.id_to_token.size()
+            );
+        }
```
Comment (slaren) [inline llama.cpp:2340]: `llama.cpp` shouldn't print directly to the console. Not sure what is the intended solution here, either use `LLAMA_LOG_*`, enclose them in `#ifdef PRETOKENIZERDEBUG`, or remove them entirely.

Comment (staviq) [reply llama.cpp:2340]: Oh, I missed that.

I'm gonna do a follow-up PR, because some kind of sanitizer function would be handy, there currently isn't a way to cleanly strip special tokens in `main` so it still prints them, so a couple of small improvements can be made. I'll include any suggestions there.

**Comment by cebtenzzre** [comment]
> I've also seen it generate broken EOS tokens (e. g. missing angle brackets)

True EOS is a single token to the model, it either generates it or it doesn't. A partial EOS token could be an attempt at a literal `</s>` (HTML strikethrough closing tag), colloquial sarcasm tagging (`/s`), or somewhere between training and inference a buggy tokenizer saw `</s>` and did not tokenize it properly.

**Comment by StefanDanielSchwarz** [comment]
Most recent cases were with models using ChatML where the `<|im_end|>` was "misspelled" as `|im_end|` and various permutations. So I guess that was because it didn't see these EOS tokens as tokens but strings and repetition penalty caused it to drop the angle bracket tokens in that case (probably had some HTML output before where it saw the brackets).

Proper tokenization would fix that, so I guess this is now a solved issue with this PR and https://github.com/ggerganov/llama.cpp/issues/3475 - but what about the repetition penalty affecting the EOS token, is that true? I think special tokens need to be exempt from repetition penalty or else we'll see such breakage again with ongoing conversations.

**Comment by staviq** [comment]
@StefanDanielSchwarz
> For lack of a better place to ask this, and considering this PR is about tokenization and special token handling:
> 
> Do special tokens get special handling, i. e. are they exempt from repetition penalty?
> 
> Because if not, with every message that's in the context and ends with an EOS token, it gets less likely (depending on repetition penalty settings) that the EOS token will be generated by the model, right? The model would be forced to continue generating until hitting max new tokens limit, and likely ramble or derail. I've also seen it generate broken EOS tokens (e. g. missing angle brackets), which could be caused by tokenizer or repetition issues.

This particular PR only allows for processing special tokens on the "input side" of the model, at the "text to tokens" stage, before any actual inference happens.

Repetition penalty happens on the "output side" of the model, at which point generated tokens have no ambiguity which would have to be processed in any special way.

Your intuition about bos/eos occurring many times in the context is actually somewhat correct, but in a different way, which depends entirely on the particular model itself.

Basically:

Chat type models are trained to understand multiple "messages" each with its own bos/eos (or equivalent special tokens), in the same context.

Instruct type models, if trained on a prompt format where user message and AI response have their own bos/eos, will understand multiple bos/eos, but if you give it consecutive prompts they might have trouble understanding whether your input is a continuation of a previous prompt, or a "reset" meaning everything prior doesn't matter. This problem depends entirely on how "smart" a given model is.

A raw model, not trained on any specific prompt format, if it has/used a bos/eos would probably start producing garbage as soon as additional bos/eos is encountered.

**Comment by ggerganov** [comment]
Fixing the repetition penalty to either not affect special tokens or a list of user-defined tokens can be implemented easily. See the `llama_sample_repetition_penalty()` function in `llama.cpp`.

**Comment by shibe2** [comment]
Are there models that can output special tokens besides EOS?

**Comment by shibe2** [comment]
@staviq Which such model can I download to play with? Using ChatML for input does not imply ability of producing same special tokens in output.

**Comment by staviq** [comment]
> @staviq Which such model can I download to play with? Using ChatML for input does not imply ability of producing same special tokens in output.

Uhmm, actually I can't reproduce that after special token handling merge, so I'm not sure now if I'm misremembering this or it was something specific I tested 3 weeks ago that I'm doing different now

Assume I was wrong about this.

**Comment by StefanDanielSchwarz** [comment]
Just FYI: There's now an issue for it on the TGI repo: https://github.com/huggingface/transformers/issues/26902

Hopefully this repetition penalty issue will be fixed for all inference software soon...

**Comment by jploski** [comment]
Interestingly enough, "<|im_end|>" isn't just an added or special token, it's also _the_ EOS token in [Mistral-OpenOrca](https://huggingface.co/Open-Orca/Mistral-7B-OpenOrca/blob/main/tokenizer_config.json)... So a multi-turn chat contains multiple EOS tokens, which by itself feels like an abuse of the concept of "EOS". But it doesn't change anything with respect to the problem of imposing an unwanted repetition penalty on such tokens.

Note that this same problem also exists in chat models that use normal tokens for separating turns (the good old "reverse prompt" parameter in llama.cpp's terms, such as "### Assistant:" - I suppose those would get repetition-penalized as well).

**Comment by StefanDanielSchwarz** [comment]
@jploski Yes, this has been a problem all this time. Glad it's getting the necessary attention now.

I used to think that the EOS token should never be part of the prompt, instead it should only be output by the model to end inference. The inference backend would catch it, stop inferencing, and remove it before sending the output to the frontend. That way it would never show up in the next prompt, and usually the `</s>` worked like that.

Now that the ChatML format is getting traction and might turn into a standard prompt template, the issue has become more apparent. But yes, the `### Assistant:` string was also affected and I've seen it get "rewritten" in various ways when repetition penalty prevented it from being printed the proper way.

Same with the LLM outputting a character's name as a prefix for its message when chatting. After a while, it would fail to do so and break the chat. The solution to that was to have the frontend manipulate the prompt and add the character name itself, so the model only completes after that and never has to write the name by itself.

**Comment by teleprint-me** [comment]
I've been running into [this issue in the templates](https://github.com/abetlen/llama-cpp-python/issues/717#issuecomment-1765167553).

I've run several tests and it's usually always the same outcome, regardless of the model.

I know this PR is adding support for special tokens, but were the special tokens handled internally? I haven't had the opportunity to dig into the code yet.

**Comment by shibe2** [comment]
I experimented with [Mistral-7B-OpenOrca](https://huggingface.co/Open-Orca/Mistral-7B-OpenOrca). It indeed outputs im_end.

**Comment by teknium1** [comment]
Yep, all chatml models made by axolotl uses <|im_end|> to seperate turns, as well as EOS token. 

I havent noticed an issue with 13~ turns when inferencing Hermes 2, but I think its mostly due to the long responses it gives in my turns, which makes the start/end/assistant/user tokens that show up each turn less disasterous. I think this effect will be most prominent in short token turns - i.e. roleplaying or a model who is trained to answer in as few words as possible type scenarios, but no doubt this issue has been affecting **all** multiturn models in some way, whether subtlely or not since at least the first sharegpt/vicuna models came around, without anyone being aware until very recently somehow, probably due to EOS tokens being throughout all turns making the problem even more exacerbated

**Comment by StefanDanielSchwarz** [comment]
Exactly. I'm one of those chatters/roleplayers who have been experiencing these issues (responses getting longer, less coherent, until eventually derailing) since the beginning (Alpaca/Vicuna).

Used to think it's just a limit of the technology and worked around it by setting various permutations of the EOS token as stopping strings. So glad this is finally getting the attention it deserves and hopefully will be fixed globally.

That should make our local/open LLMs much more intelligent and useful once they work as intended and no longer hallucinate/derail that much. Especially the smaller models that tend to require closer adherence to the trained/tuned prompt format than bigger models which tend to cope better (but even they should profit).

**Comment by jploski** [comment]
> That should make our local/open LLMs much more intelligent and useful once they work as intended and no longer hallucinate/derail that much.

I doubt it will have that much effect because if repetition penalty messing up generation was the issue, we should also be seeing it for newline tokens (much more frequent than EOS / next turn tokens). The derailment is more likely due to paying equal attention to their own (garbage) outputs as anything else and eventually sampling more and more out of training distribution as a result. So yes, I'd say it is an inherent limitation of the technology. Hallucinations are also out-of-distribution sampling (to prevent you'd either have to specifically train on such garbage samples or maybe detect occurrences, maybe by scoring semantical proximity of multiple outputs of the same prompt; e.g. if the model is happy to generate "yes" and "no" to the same question with same probability, it's safe to infer it does not know the answer).

As for the repetition penalty, it is not really all that much needed if you sample from the entire distribution (high top_p) rather than greedily. The reason why repetitions happen with greedy sampling: imagine you have a biased coin (your LLM) and with each prediction you only pick the most probable outcome. With such a sampling method you'd only ever get heads or tails, but never a realistic mix of them generated by such a coin. (Note: top_k can also be very greedy if all the top k tokens are synonyms. This description also generalizes to multi-token generation, imagine throwing n different biased coins; then you get the GPT-style sentence-level loops). Finally, the repetition seems to also have someting to do with the amplitude swings in attention scores in the positional embeddings (both original GPT and RoPE).. based on anecdotal evidence that in a model with different positional embeddings (RetNet + XPos), repetitions also look different (individual token stutter - but followed by auto-recovery, no endless self-reinforcing looping).

**Comment by StefanDanielSchwarz** [comment]
Very insightful post! Just to add my observations:

> we should also be seeing it for newline tokens (much more frequent than EOS / next turn tokens)

I think that's also visible. I used examples split into multiple paragraphs, and the splitting became inconsistent, probably because the newline became penalized. Same with other characters like dashes, quotes, asterisks - when used a lot, the model would suddenly switch, using different unicode symbols for dashes, quotes, or even switch from `*this is an action*` to `(this is an action)` or `[this is an action]` formats.

I think we need a better solution than simple (stupid) repetition penalty. There's also relevant discussion here: https://github.com/huggingface/transformers/issues/26902

File: llama.cpp [hunk 6]
```diff
@@ -2241,6 +2244,101 @@ static void llm_load_vocab(
     GGUF_GET_KEY(ctx, vocab.special_unk_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_UNK_ID));
     GGUF_GET_KEY(ctx, vocab.special_sep_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_SEP_ID));
     GGUF_GET_KEY(ctx, vocab.special_pad_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_PAD_ID));
+
+    // build special tokens cache
+    {
+        // TODO: It is unclear (to me) at this point, whether special tokes are guaranteed to be of a deterministic type,
+        //  and will always be correctly labeled in 'added_tokens.json' etc.
```
Comment (ArthurZucker) [inline llama.cpp:2251]: Sorry if it's irrelevant, but the `added_tokens.json` indeed does not have types, however with the recent update, we save the `added_tokens_decoder` which includes everything. See the 2 prs [here](https://github.com/huggingface/transformers/pull/23909)  and [here](https://github.com/huggingface/transformers/pull/26570) in `transformers`

Comment (staviq) [reply llama.cpp:2251]: I might have worded that code comment a bit wrong, I didn't mean `added_tokens.json` in particular, but rather source model metadata and the way it's processed by import scripts in general.

BPE models are still problematic, at least couple of them I tested this with, had no token types imported/defined at all. So my suspicions were correct, and at this point in time that TODO is still reasonably valid.

Though the method used in this PR to detect special tokens, was/is meant to only be used as a fallback, eventually.

**Comment by cebtenzzre** [comment]
> I played a bit with hardcoded space prefix always vs only for user input itself, and I'm getting subjectively much much better results, if the hardcoded space is only added to the user input itself and not the "prompt template" ( prefix and suffix ).

@staviq Are we sure this was the right call? We should look into what HF transformers does.

Command: `build/bin/main -t 16 -m /tmp/dolphin-llama2-7b.Q4_0.new.gguf -n 2048 --ignore-eos -p 'The quick brown fox' --seed 1699464025`

Before this PR (with a leading space, GitHub doesn't show it):
> The quick brown fox jumps over the lazy dog is a well-known English nursery rhyme or children's song.➖
>
> This nursery rhyme has been popular since 1847, when it was first published in the Monthly Magazine of London. Since then, it has undergone various changes and adaptations, but its essence remains the same. The rhyme tells the story of a quick brown fox who jumps over the lazy dog to avoid being caught for stealing some cheese.

After this PR (no leading space):
> The quick brown fox jumps over the wall (the last line of the poem) and is on the other side.nahmáo Often used as a verb, němäo is a term derived from the Sanskrit word nirvana, meaning "spirit" or "consciousness." The concept of the spirit or consciousness can be found in many cultures and religions throughout history. In Buddhism, for instance, the term "nirvana" refers to the state of perfect spiritual awakening that is the goal of practice. Similarly, in Christianity, the Holy Spirit is considered a divine spirit that guides and supports believers.

**Comment by staviq** [comment]
@cebtenzzre 

The problem was that making special tokens work, allowed proper prompt formats to also work, except when the space prefix was added to `-p`, something like `SYSTEM:` still wasn't working properly because ` SYSTEM` wasnt correctly triggering model's prompt format, but `SYSTEM` was.

So it was a tradeoff between making prompt formats work better and making `main` more complicated by introducing another bunch of parameters, for splitting `-p` into prefix, content and suffix, each processed with/without special tokens or space prefix.

There was some talk about splitting/changing `add_bos`, `special` flags for tokenizer, and separate one for space prefix, but I don't think that PR went through, or I missed it.

Feel free to revert it, if it's causing problems.

I can't think if a way to determine whether `-p` is or isn't using some sort of prompt format, to decide automatically if or where should a space prefix be added.

**Comment by shibe2** [comment]
If a space is desirable but missing, the user can add it. If a space is added automatically but is undesirable, the user is out of luck.

**Comment by cebtenzzre** [comment]
@staviq
Models with GPT2-style BPE have add_prefix_space in tokenizer.json/tokenizer_config.json, I believe we need to honor that.

As for llama, this is what HF transformers does (I assume we should match the legacy=False behavior):
```python
"""
Converts a string to a list of tokens. If `self.legacy` is set to `False`, a prefix token is added unless the
first token is special.
"""
```
```python
if len(tokens) > 1 and tokens[0] == SPIECE_UNDERLINE and tokens[1] in self.all_special_tokens:
    tokens = tokens[1:]
```

And an interesting note about add_dummy_prefix (changed in [this PR](https://github.com/huggingface/transformers/pull/25224)):
```python
"""
Returns a tokenized string.

We de-activated the `add_dummy_prefix` option, thus the sentencepiece internals will always strip any
SPIECE_UNDERLINE. For example: `self.sp_model.encode(f"{SPIECE_UNDERLINE}Hey", out_type = str)` will give
`['H', 'e', 'y']` instead of `['▁He', 'y']`. Thus we always encode `f"{unk_token}text"` and strip the
`unk_token`. Here is an example with `unk_token = "<unk>"` and `unk_token_length = 4`.
`self.tokenizer.sp_model.encode("<unk> Hey", out_type = str)[4:]`.
"""
```

<br>
I think it is important to fix this regression for models that do not use special tokens in the prompt (including all of the ones that I use).

**Comment by shibe2** [comment]
If you don't want to add space after special tokens, only add it to a fragment if it precedes all special tokens except BOS, i.e. on the first iteration over `fragment_buffer`. This way, prompts without special tokens will get space as before.

**Comment by ArthurZucker** [comment]
> If a space is desirable but missing, the user can add it. If a space is added automatically but is undesirable, the user is out of luck.

Hey 🤗  The `add_prefix_space` parameter will be added for `llama`-like tokenizers (both slow and fast) soon-ish.
The `legacy` is related to the space after special tokens that was always added due to the underlying `add_dummy_prefix_space` of the sentencepiece model.

**Comment by shibe2** [comment]
>The `add_prefix_space` parameter will be added for `llama`-like tokenizers (both slow and fast) soon-ish.

Is there any documentation about that?
